libpod/container.go

```diff
@@ -1,76 +1,312 @@
+	RootfsDir string `json:"rootfsDir,omitempty"`
+	// Information on the image used for the root filesystem
+	RootfsImageID   string `json:"rootfsImageID,omitempty"`
+	RootfsImageName string `json:"rootfsImageName,omitempty"`
+	UseImageConfig  bool   `json:"useImageConfig"`
+	// Whether to keep container STDIN open
+	Stdin bool
+	// Static directory for container content that will persist across
+	// reboot
+	StaticDir string `json:"staticDir"`
+	// Pod the container belongs to
+	Pod *string `json:"pod,omitempty"`
+	// Shared namespaces with container
+	SharedNamespaceCtr *string           `json:"shareNamespacesWith,omitempty"`
+	SharedNamespaceMap map[string]string `json:"sharedNamespaces"`
+	// Time container was created
+	CreatedTime time.Time `json:"createdTime"`
+
+	// TODO save log location here and pass into OCI code
+	// TODO allow overriding of log path
 }
 
 // ID returns the container's ID
 func (c *Container) ID() string {
-	// No locking needed, ID will never mutate after a container is created
-	return c.id
+	return c.config.ID
 }
 
 // Name returns the container's name
 func (c *Container) Name() string {
-	return c.name
+	return c.config.Name
 }
 
 // Spec returns the container's OCI runtime spec
+// The spec returned is the one used to create the container. The running
+// spec may differ slightly as mounts are added based on the image
 func (c *Container) Spec() *spec.Spec {
-	// The spec can potentially be altered when storage is configured and to
-	// add annotations at container create time
-	// As such, access to it is locked
-	c.lock.RLock()
-	defer c.lock.RUnlock()
-
 	spec := new(spec.Spec)
-	deepcopier.Copy(c.spec).To(spec)
+	deepcopier.Copy(c.config.Spec).To(spec)
 
 	return spec
 }
 
+// State returns the current state of the container
+func (c *Container) State() (ContainerState, error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// TODO uncomment when working
+	// if err := c.runtime.ociRuntime.updateContainerStatus(c); err != nil {
+	// 	return ContainerStateUnknown, err
+	// }
+
+	return c.state.State, nil
+}
+
+// The path to the container's root filesystem - where the OCI spec will be
+// placed, amongst other things
+func (c *Container) bundlePath() string {
+	return c.state.RunDir
+}
+
+// Retrieves the path of the container's attach socket
+func (c *Container) attachSocketPath() string {
+	return filepath.Join(c.runtime.ociRuntime.socketsDir, c.ID(), "attach")
+}
+
 // Make a new container
 func newContainer(rspec *spec.Spec) (*Container, error) {
 	if rspec == nil {
 		return nil, errors.Wrapf(ErrInvalidArg, "must provide a valid runtime spec to create container")
 	}
 
 	ctr := new(Container)
-	ctr.id = stringid.GenerateNonCryptoID()
-	ctr.name = ctr.id // TODO generate unique human-readable names
+	ctr.config = new(containerConfig)
+	ctr.state = new(containerRuntimeInfo)
+
+	ctr.config.ID = stringid.GenerateNonCryptoID()
+	ctr.config.Name = ctr.config.ID // TODO generate unique human-readable names
```

Comment:
Safe to assume withName overrides this ?

Reply:
Yeah, WithName will run after newContainer and change whatever name is recorded here